staging/src/k8s.io/kube-apiextensions-server/pkg/controller/status/naming_controller.go

```diff
@@ -0,0 +1,312 @@
+// This could eventually be lifted, but starting simple.
+type NamingConditionController struct {
+	customResourceDefinitionClient client.CustomResourceDefinitionsGetter
+
+	customResourceDefinitionLister listers.CustomResourceDefinitionLister
+	customResourceDefinitionSynced cache.InformerSynced
+
+	// To allow injection for testing.
+	syncFn func(key string) error
+
+	queue workqueue.RateLimitingInterface
+}
+
+func NewNamingConditionController(
+	customResourceDefinitionInformer informers.CustomResourceDefinitionInformer,
+	customResourceDefinitionClient client.CustomResourceDefinitionsGetter,
+) *NamingConditionController {
+	c := &NamingConditionController{
+		customResourceDefinitionClient: customResourceDefinitionClient,
+		customResourceDefinitionLister: customResourceDefinitionInformer.Lister(),
+		customResourceDefinitionSynced: customResourceDefinitionInformer.Informer().HasSynced,
+		queue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "CustomResourceDefinition-NamingConditionController"),
+	}
+
+	customResourceDefinitionInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.addCustomResourceDefinition,
+		UpdateFunc: c.updateCustomResourceDefinition,
+		DeleteFunc: c.deleteCustomResourceDefinition,
+	})
+
+	c.syncFn = c.sync
+
+	return c
+}
+
+func (c *NamingConditionController) getNamesForGroup(group string) (allResources sets.String, allKinds sets.String) {
+	allResources = sets.String{}
+	allKinds = sets.String{}
+
+	list, err := c.customResourceDefinitionLister.List(labels.Everything())
+	if err != nil {
+		panic(err)
+	}
+
+	for _, curr := range list {
+		if curr.Spec.Group != group {
+			continue
+		}
+
+		allResources.Insert(curr.Status.AcceptedNames.Plural)
+		allResources.Insert(curr.Status.AcceptedNames.Singular)
+		allResources.Insert(curr.Status.AcceptedNames.ShortNames...)
+
+		allKinds.Insert(curr.Status.AcceptedNames.Kind)
+		allKinds.Insert(curr.Status.AcceptedNames.ListKind)
+	}
+
+	return allResources, allKinds
+}
+
+func (c *NamingConditionController) calculateNames(in *apiextensions.CustomResourceDefinition) (apiextensions.CustomResourceDefinitionNames, apiextensions.CustomResourceDefinitionCondition) {
+	// Get the names that have already been claimed
+	allResources, allKinds := c.getNamesForGroup(in.Spec.Group)
+
+	condition := apiextensions.CustomResourceDefinitionCondition{
+		Type:   apiextensions.NameConflict,
+		Status: apiextensions.ConditionUnknown,
+	}
+
+	requestedNames := in.Spec.Names
+	acceptedNames := in.Status.AcceptedNames
+	newNames := in.Status.AcceptedNames
+
+	// Check each name for mismatches.  If there's a mismatch between spec and status, then try to deconflict.
+	// Continue on errors so that the status is the best match possible
+	if err := checkName(requestedNames.Plural, acceptedNames.Plural, allResources); err != nil {
```

Comment:
> Where is it checked that singular, plural and shortnames do not overlap? 

Inside of a single resource, they should be able to since the parsing would still be unambiguous for intent.  Think `endpoints` and `endpoints`.  They point to the same spot.